proposals/0000-typed-metadata.md

```diff
@@ -0,0 +1,577 @@
+```
+
+&nbsp;
+&nbsp;
+
+## Haxe API Changes
+
+A new optional field should be added to `haxe.macro.Expr.MetadataEntry`.
+
+If this metadata entry is typed, then `field` will contain a reference to the `ClassField` of the metadata function.
+```haxe
+// Unresolved question
+// Would it be possible to use Ref<Type.ClassField> instead?
+var ?field: Expr.Field;
+```
+
+&nbsp;
+
+### Reading Arguments
+
+There needs to be a mechanism for reading metadata arguments. To provide this, a new field `typedMeta: StringMap<Dynamic>` should be added to:
+ * `haxe.macro.Expr.TypeDefinition`
+ * `haxe.macro.Expr.Field`
+ * `haxe.macro.Expr.TypeParamDecl`
+
+The entires correlate directly to the full path of the metadata used on the subject. So to access the content of an `@Meta.date` metadata, `_.typedMeta.get("mypack.Meta.date")` must be used. This is to prevent naming conflicts. There may be multiple metadata of the same name in different modules.
+
+The `Dynamic` value contains fields with the same name as the arguments of the typed metadata. These fields store the values passed to the metadata entry. How these values are converted can be viewed in [Allowed Argument Types](0000-typed-metadata.md#allowed-argument-types).
+
+If the metadata has `allowMultiple` enabled, the `Dynamic` value will ALWAYS be an Array, even if only one instance of the metadata is used.
+```haxe
+// MyModule.hx
+package mypack;
+
+@:metadata({ allowMultiple: true })
+function author(name: String): TypeDefinition;
+
+class Meta {
+   @:metadata
+   public static function date(month: Int, day: Int): TypeDefinition;
+}
+class AnotherMeta {
+   @:metadata
+   public static function date(dateString: String): TypeDefinition;
+}
+
+@author("Something")
+@Meta.date(11, 15)
+@AnotherMeta.date("November 15, 2004")
+class MyClass {}
+```
+
+```haxe
+// ---
+// in some compile-time function
+// var td: TypeDefinition;
+final authorNames: Null<Array<String>> = td.typedMeta.get("mypack.MyModule.author")?.map(meta -> meta.name);
+
+final dateMonth: Int = td.typedMeta.get("mypack.MyModule.Meta.date")?.month;
+```
+
+&nbsp;
+
+### Context.typeMetadata
+
+A new static function should be added to `haxe.macro.Context`:
+```haxe
+class Context {
+    // ...
+    public static function typeMetadata(meta: haxe.macro.Expr.Metadata): StringMap<Dynamic> { ... }
+```
+
+This is a function that will generate an object like the `typedMeta: StringMap<Dynamic>` field described in the previous section. This would be helpful for extracting typed metadata data found in untyped `EMeta` expressions.
+
+&nbsp;
+
+### Function Type Struct
+
+The following anonymous structure should be added to the `haxe/macro/Expr.hx` module:
+```haxe
+typedef FunctionPath = {
+	> TypePath,
+	functionName: String;
+};
+```
+
+This is a structure for storing type paths to functions. It is used as an argument type for metadata. Long story short, it allows for type paths that end with a lowercase identifier (`myFunc`, `Module.Sub.myFunc`).
```

Comment:
Why should that be function-specific? I'd expect it to be more like 
```haxe
typedef FieldPath = {
	> TypePath,
	field: String;
};
```

Reply:
That's a great point, oversight on my part.